Simplify ParentEntityEditPanel

labkey-danield opened the pull request:
#### Rationale
Simplify the ParentEntityEditPanel. When this panel was first implemented it was possible to have multiple edit panels open on a page, that is no longer possible. This simplifies the way the edit panel is found, and the methods that are needed.

Example of Adding a Parent:
![Screen Shot 2022-03-22 at 9 25 22 PM](https://user-images.githubusercontent.com/12738200/159623808-ec6d977c-9d4c-40d3-88a3-1b59b862aae6.png)


Example of editing with multiple parent types
![Screen Shot 2022-03-22 at 9 22 18 PM](https://user-images.githubusercontent.com/12738200/159623453-2857ba65-173d-4d6f-bc24-df961066556e.png)


#### Related Pull Requests
* https://github.com/LabKey/sampleManagement/pull/887

#### Changes
* Removed the need to pass in a type enum to find the edit panel.
* Added a finder.
* Updated calls to the reactSelect finders to use the hidden text box (the name of the type is in the name of the input).
* Clarified method names to differentiate between the select for the type (sample or source type name) and the parent (the specific sample or source to be used as the parent).
* Updated comments.

labkey-danield committed:
Check that the save or cancel button is enabled before clicking it.

labkey-danield committed:
Refactored ParentEntityEditPanel to be entity type agnostic.
Removed redundant panel objects and enums.
Updated SampleOverviewPage to check for readiness of edit panel before returning it.

labkey-danield committed:
Merge branch 'develop' into fb_waitForSaveButtonEnabled

labkey-danield committed:
Merge branch 'develop' into fb_waitForSaveButtonEnabled

labkey-danield committed:
Increase wait time for panel to change state when updating lineage.

labkey-danield committed:
Merge branch 'develop' into fb_waitForSaveButtonEnabled

labkey-tchad reviewed (changes requested):
This component should be able to wait for itself to be ready. The difference between the initial state with zero parents defined and editing existing parents is that there is no "Parent IDs" select when the panel loads with no existing parents.
We should be able to wait for at least one parent type select to be present and for any "Parent ID" selects to have at least one selection loaded.

labkey-danield committed:
Merge branch 'develop' into fb_waitForSaveButtonEnabled

labkey-danield committed:
In BaseReactSelect check that the element to be removed is no longer visible, not just stale.
Overloading the waitForReady method so the edit panel can check for itself if it is ready to go.
Using a better sample for the test.

labkey-danield committed:
Reverting change in BaseReactSelect.
Adding a pause in the test to make sure selections are removed from the FilteredReactSelect control.

labkey-danield committed:
Implementing some SonarLint suggestions.
Simplifying check to see if there are no parents yet.

labkey-danield committed:
Changed list check to use isEmpty.
Renamed variable to hopefully be a little more meaningful.

labkey-danield commented:
I believe that I have the form able to identify when it is ready.

labkey-tchad commented on src/org/labkey/test/components/ui/entities/ParentEntityEditPanel.java:
```diff
@@ -8,80 +8,114 @@
+            {
+
+                // Check to see if there are no existing parents.
+                List<ReactSelect> firstParentSelector = ReactSelect.finder(getDriver())
+                        .withNamedInput("entityType0")
+                        .findAll(this);
+
+                Assert.assertFalse("It looks like there are multiple new parent ReactSelect controls present, that should never happen.",
+                        firstParentSelector.size() > 1);
+
+                if(!firstParentSelector.isEmpty())
+                {
+                    return firstParentSelector.get(0).isInteractive();
+                }
```
I think you want to check the state of the entityType selector second. `firstParentSelector` should never never be empty, so this will always return before checking for any selected source/parent IDs.

labkey-danield replied:
Ok, hopefully third time is the charm.

labkey-danield committed:
Improve check for panel ready.

labkey-tchad commented on src/org/labkey/test/components/ui/entities/ParentEntityEditPanel.java:
```diff
@@ -8,80 +8,114 @@
+                    {
+                        for (FilteringReactSelect select : filteringReactSelects)
+                        {
+                            if (!select.isInteractive())
```
Should wait for selections as well.
```suggestion
                            if (!select.isInteractive() || !select.hasSelection())
```

labkey-tchad reviewed (approved):
Approved pending previous suggestion.
